keps/sig-api-machinery/0034-20190129-kflag.md

```diff
@@ -0,0 +1,188 @@
+
+#### Application of Parsed Flag Values
+* Components must be able to apply the parsed values an arbitrary number of times without
+  accumulating side-effects. Components may apply values once for at least as many config sources
+  they can consume (at least once to access the initial flag values, and afterward to enforce flag
+  precedence on each config). Each stage of loading configuration (flags, config file, dynamic
+  config) is a potential decision point, prior to which flag precedence must be enforced.
+* Components should be able to apply the parsed flag values to structs that only represent a subset
+  of all flag values. For instance, if a component has a Flags struct and a Config struct,
+  it may need to apply values to each independently.
+* Components must be able to specify custom behavior for merging values during application.
+  Feature gates, for example, are merged piecemeal between the command-line and config files, with
+  gates specified on the command-line taking precedence.
+* Application of flag values should only consider the flags specified on the command line. It should
+  be careful, for example, to not overwrite values loaded from config files with the default values
+  of omitted flags.
+
+
+### Implementation Details
+
+Fundamentally, separating parsing from application requires that we have a scratch space to parse
+the initial values into, and a way to map that scratch space onto target objects.
+
+To maintain type safety, this proposal recommends a two-layered approach, where kflag handles the
+common types, and components handle the component-specific types:
+* kflag Implements:
+  * Statically typed flag registration helpers that allocate scratch space.
+  * Statically typed helpers to apply the values of parsed flags to arbitrary targets.
+* Components Implement:
+  * Aggregate flag registration functions, which use kflag helpers to register groups of flags the
+    component cares about.
+  * Aggregate flag application functions, which use kflag helpers to apply a subset of flag values
+    to an arbitrary structure determined by the component.
+
+This proposal recommends the approach in this example PR:
+* https://github.com/kubernetes/kubernetes/pull/73494
+
+See also @sttts's prototype for an alternative approach that predates this KEP:
+* https://github.com/kubernetes/kubernetes/pull/72037
+
+### Risks and Mitigations
+
+#### Maintenance
+The kflag package is a wrapper around an existing package, pflag.
+The pflag package has a stable API, so we would at most expect to extend kflag when pflag is
+extended, but only in the event that we really need that extension. We expect this to be infrequent.
+Given that kflag should make components easier to maintain (it reduces the complexity of the
+bootstrap process by making the flag precedence implementation easier to use and reason about),
+the maintenance of kflag itself is likely time well spent.
```

Comment:
kflag should also be able to hot fix bugs in pflag package.

Reply:
> kflag should also be able to hot fix bugs in pflag package.

depends if it's doable given kflag only wraps pflag (from my undertanding).
if a core problem emerges in pflag a patch upstream or a patch in a fork is needed.

Reply:
The intent is not to fork pflag, but to provide a relatively thin wrapper around it. It may or may not be possible to plaster over bugs with this wrapper.